### PR TITLE
Don't show deleted quick link pages, refs #13297

### PR DESCRIPTION
**Pull request by mcantelon**
Added logic to the quick links menu so links won't be shown to any static pages that have been deleted.

### CODE REVIEW

File: apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php [hunk 1]
```diff
@@ -28,12 +28,26 @@ class menuQuickLinksMenuComponent extends sfComponent
...
   {
-    // Get menu objects
-    $this->quickLinksMenu = QubitMenu::getById(QubitMenu::QUICK_LINKS_ID);
+    // Get menu
+    $quickLinksMenu = QubitMenu::getById(QubitMenu::QUICK_LINKS_ID);
 
-    if (!$this->quickLinksMenu instanceof QubitMenu)
+    if (!$quickLinksMenu instanceof QubitMenu)
     {
       return sfView::NONE;
```
Comment (djjuhasz) [inline apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php]: Nitpick: I believe `return sfView::NONE` is only relevant for sfAction objects, and a bare `return;` is sufficient to skip component output.

File: lib/Qubit.class.php [hunk 2]
```diff
@@ -452,4 +452,13 @@ public static function generateIdentifierFromCounterAndMask($counter, $mask)
       }
     }, $mask);
   }
+
+  public static function actionExistsForUrl($url)
```
Comment (djjuhasz) [inline lib/Qubit.class.php]: This method looks like a duplicate of the `QubitObject::actionExistsForUrl()` method, and it doesn't look like it's called anywhere.

File: apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php [hunk 3]
```diff
@@ -28,12 +28,26 @@ class menuQuickLinksMenuComponent extends sfComponent
...
+
+    // Get menu items that correspond to an external URL or an internal path
+    $this->quickLinks = [];
+
+    foreach ($quickLinksMenu->getChildren() as $child)
+    {
+      $url = $child->getPath(['getUrl' => true, 'resolveAlias' => true]);
+      $urlParsed = parse_url($url);
+
+      if (isset($urlParsed['scheme']) || QubitObject::actionExistsForUrl($url))
```
Comment (djjuhasz) [inline apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php:47]: There's a bug in this logic somewhere - `$this->quickLinks` is always empty in my test AtoM instance, even though I have quick link menu items.

empty_quick_links_menu

**Review by djjuhasz — CHANGES_REQUESTED** [review]
@mcantelon I think there is a bug in the logic here somewhere - my "Quick links" menu is always empty when I check out this dev branch.

File: lib/Qubit.class.php (hunk 2 above)
Comment (mcantelon) [reply lib/Qubit.class.php]: Oops... accidentally committed that.

File: apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php (hunk 1 above)
Comment (mcantelon) [reply apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php]: Fixed!

**Comment by mcantelon** [comment]
@djjuhasz I tried this branch in another dev VM and it works for me. Maybe restart your VM in case there's some stubborn caching going on?

File: apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php (hunk 3 above)
Comment (djjuhasz) [reply apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php:47]: I retested, and the quick links are working for me now.

**Review by djjuhasz — APPROVED** [review]
@mcantelon you are right, I restarted my test VM and the quick links are working now. 🤷 

The code looks good to me.  👍 

I am curious about the added check to make sure the menu path is a valid URL or internal route.   I doesn't mention anything in the ticket about checking the paths. Is the path checking logic something you thought would be good to do, or was there another reason for it?

**Comment by mcantelon** [comment]
@djjuhasz The path checking is a means of checking to make sure a menu item doesn't point to a static page, or any other AtoM resource/action, that has been deleted.

If the path is an absolute URL then it's deemed external and there's no check done to see if the external resource exists. If the path isn't an absolute URL then it's treated as a path to a Symfony resource/action. If a static page has been deleted then the menu item's path (a static page slug) won't resolve to a Symfony action and the menu item ends up getting skipped and not displayed.

**Comment by djjuhasz** [comment]
@mcantelon oh, so instead of deleting the menu item to stop e.g. the privacy policy page from showing in the quick links menus one would delete the static page?  🤔   From the ticket description I was just expecting a "delete" button on the menu item edit page.

**Comment by mcantelon** [comment]
@djjuhasz Ah, you're right I misread the issue! I'll fix the inability to delete the menu item itself and add that to this PR.

**Comment by mcantelon** [comment]
@djjuhasz I've added the ability to delete the privacy menu item. Let me know if you see anything that might need to be tweaked (I'll update the commit message, before merging. to reflect the DB tweak to unprotect the privacy menu item).Thanks!

**Review by djjuhasz — APPROVED** [review]
Looks good to me @mcantelon. 😄

**Comment by mcantelon** [comment]
Woot! Thanks @djjuhasz !

**Comment by Jimadine** [comment]
@mcantelon I upgraded from `2.6.4` to `2.7.1` and noticed that our custom quick links are missing from the quick links menu. These are links with the path set to `staticpage/index?slug=<slug>` (external `http`/`https` links unaffected).  I tried a VM restart as suggested earlier in this PR but no luck. So I edited `apps/qubit/modules/menu/actions/quickLinksMenuComponent.class.php`, commenting out the conditional as follows:
```
            //if (isset($urlParsed['scheme']) || QubitObject::actionExistsForUrl($url)) {
                $this->quickLinks[] = $child;
            //}
```
The static pages definitely exist and have the slug field set.
With the above conditional commented out, after running `php symfony cc` & `systemctl restart php7.4-fpm`, the quick links appeared.
I've also checked this in a clean VM with `stable/2.7.x` to rule out any local customisations, and the problem is still apparent. Are you able to take a look, please? An example missing link is to the privacy page.

**Pull request merged**